File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
    void getSubsets(vector<int>& candidates, int i, int N, int curSum, int target, vector<int>& curSubset, vector<vector<int>>& subsets) {
        
        if(curSum == target) {
            subsets.push_back(curSubset);
            return;
        }
        if(curSum > target) return;
        if(i == N) return;
        
        curSubset.push_back(candidates[i]);
        getSubsets(candidates, i +1, N, curSum +candidates[i], target, curSubset, subsets);
        
        curSubset.pop_back();
        while(i +1 < N && candidates[i] == candidates[i +1]) i++;
        getSubsets(candidates, i +1, N, curSum, target, curSubset, subsets);
    }
    
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        
        vector<vector<int>> subsets;
        vector<int> curSubset;
        
        getSubsets(candidates, 0, candidates.size(), 0, target, curSubset, subsets);
        
        return subsets;        
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> combinations = new ArrayList<>();
        generateCombinations(candidates, 0, target, new ArrayList<>(), combinations);
        
        return combinations;
    }
    
    private void generateCombinations(int[] candidates, int pos, int rem, 
                                      List<Integer> currCombination, 
                                      List<List<Integer>> combinations) {
        
        if(rem == 0) {
            combinations.add(new ArrayList<>(currCombination));
            return;
        }
        if(rem < 0 || pos >= candidates.length) return;
        
        //take
        currCombination.add(candidates[pos]);
        generateCombinations(candidates, pos +1, rem -candidates[pos], currCombination, combinations);
        currCombination.remove(currCombination.size() -1);
        
        //skip
        while(pos +1 < candidates.length && candidates[pos] == candidates[pos +1]) pos++;
        generateCombinations(candidates, pos +1, rem, currCombination, combinations);
    }
    
}